### ISSUE TITLE
Make sure the content audits still work

### ISSUE THREAD

**Issue by smashwilson** [deconst/strider-deconst-content#5]
@ktbartholomew has been working on writing plugins that do additional lints and checks during the preparation process. I want to make sure that these still work in the automated builds.

It might be worth it to tackle deconst/deconst-docs#133 and rework the way preparers are run entirely, including a phase for 0 to n linter plugins to run somehow. Maybe something like:
1. Run the preparer container with a volume mount. Rather than submit anything to the content store, It just writes `#{content_id}.json` files somewhere within the volume and communicates back the output path. If the preparer fails, report a failed build.
2. Run 0..n linter plugin containers, mounting the same volume and linting all `.json` files found at the output path. After all linters complete, if any have failed, report a failed build with the details.
3. Run a "submitter" container that uses the API key to shunt all of the `.json` files up to the content store API. It can be clever and tar/gzip them up first for Mad Performance Gains.

This would actually let us centralize the content store communication and API key wrangling, which would be pretty nice. I haven't looked at the existing linter stuff at all though, so this may not be sufficient.

**Comment by ktbartholomew** [comment]
@smashwilson [jekyll-audit](https://github.com/ktbartholomew/jekyll-audit) actually parses the markdown/YAML source to do its checks, so it would need access to that content, not the JSON content envelopes. [ktbartholomew/link-checker](https://github.com/ktbartholomew/link-checker) will continue to be a problem, because it needs to have an entire site available via HTTP to be useful. It is also likely to induce heavy load and take a very long time to run, so may be best as an out-of-band process that doesn't hold up the rest of the build. I digress...

Generally though, I would think that steps 1 and 2 as you've suggested should be reversed. I think it would be much easier to reason about the content in its original format, and only preparing it into json envelopes if all the linters pass.

**Comment by smashwilson** [comment]
Doesn't that require you to rewrite the linters for every preparer, though?

**Comment by smashwilson** [comment]
Yeah, and the link checker feels like a separate process we could run on a regular schedule elsewhere.

**Comment by ktbartholomew** [comment]
> Doesn't that require you to rewrite the linters for every preparer, though?

What you're suggesting would require me to rewrite the one linter I already have, anyways. :stuck_out_tongue_closed_eyes: I think it makes a little more sense to reason about the content in its original format, rather how it appears _after_ it's been tinkered with by the preparer. Also, in the interest of open-source :heart: I'd like the linter to be of some utility to all other Jekyll users, not just those who are using deconst.

**Comment by smashwilson** [comment]
Hah. This issue is just for me to not lose sight of this, don't go rewriting everything yet! I'll most likely tackle the infra bits before I get back to this. I won't want to :shipit: all to production without this but we're still a fair distance from that anyway.

We can always do both: have some in-preparer linters that operate as plugins for Jekyll or Sphinx, which are just installed and bundled directly within the preparer container, and others that operate on the prepared JSON. (I'm thinking of simple grammar checks, duplicate-word finders, things like that.)
